### PR TITLE
#710 - remove support for v0

### LINKED ISSUE #710: Drop support for v0
This allows us to remove any code related to Custom Elements / Shadow DOM V0, dropping a bit of weight and code to normalise behaviour for Custom Elements.

### PR DESCRIPTION
**Pull request by treshugart**
Fixes #710.

Since v1 is support in Chrome, there's no need to support v0 even if Opera hasn't updated yet. Until they pick up the new Blink (if they haven't already) people wanting to support opera can use the polyfills.

### CODE REVIEW

File: src/api/component.js [hunk 1]
```diff
@@ -24,16 +21,10 @@ import syncPropToAttr from '../util/sync-prop-to-attr';
...
+// Uses Shadow DOM only if it's available. Once all browsers support it
+// natively then we can remove this function and call it directly.
 function attachShadow (elem) {
-  if (shadowDomV1) {
-    return elem.attachShadow({ mode: 'open' });
-  } else if (shadowDomV0) {
-    return elem.createShadowRoot();
-  }
-  return elem;
+  return elem.attachShadow ? elem.attachShadow({ mode: 'open' }) : elem;
```
Comment (zzarcon) [inline src/api/component.js]: Shouldn't `attachShadow` be always available now?

Comment (zzarcon) [inline src/api/component.js]: Oh that's true, forgot my comment then! ;)

**Comment by nadiam84** [comment]
`test/unit/api/define.js` this file also has references to the v0 custom elements.

**Comment by treshugart** [comment]
@nadiam84 I removed those references. Thanks!

File: README.md [hunk 2]
```diff
@@ -214,9 +214,9 @@ Or you can use script tags:
 <script src="https://unpkg.com/skatejs/dist/index-with-deps.min.js"></script>
 ```
 
-If you want finer grained control about which polyfills you use, you'll have to BYO Custom Element and Shadow DOM V1 polyfills. Skate will work without Shadow DOM support, but you won't be able to compose components together due to the lack of DOM encapsulation that Shadow DOM gives you.
+If you want finer grained control about which polyfills you use, you'll have to BYO Custom Element and Shadow DOM polyfills.
 
-**Skate works with the native V0 APIs, however, the V0 polyfills have several bugs that are inconsistent with the native APIs, so it's not recommended you try and use the V0 polyfills.**
+*Skate will work without Shadow DOM support, but you won't be able to compose components together due to the lack of DOM encapsulation that Shadow DOM gives you.*
```
Comment (bradleyayers) [inline README.md]: I wouldn't document this, as it makes it API. I would say

> Skate requires Custom Elements v1 and Shadow DOM v1 support. Use the skatejs-web-components library to polyfill browsers lacking native support.

File: src/api/define.js [hunk 3]
```diff
@@ -107,42 +106,22 @@ function prepareForRegistration (name, Ctor) {
...
     throw new Error(`You have to define options to register a component ${name}`);
   }
   const Ctor = typeof opts === 'object' ? Component.extend(opts) : opts;
   formatLinkedAttributes(Ctor);
 
-  if (customElementsV1) {
-    return registerV1Element(name, Ctor);
-  } else if (customElementsV0) {
-    return registerV0Element(name, Ctor);
+  if (!window.customElements) {
```
Comment (bradleyayers) [inline src/api/define.js]: I'd move this up to the top where the other assertions are made.

File: test/unit/vdom/shadow-dom.js [hunk 4]
```diff
@@ -1,34 +1,23 @@
...
-      expect(elem[symbols.shadowRoot]).not.to.equal(elem);
-    });
-  }
-
-  if (shadowDomV1) {
-    it('should work for attachShadow()', () => {
-      const elem = new Elem();
+  it('should work for attachShadow()', () => {
+    const elem = new Elem();
+    if (elem.attachShadow) {
```
Comment (bradleyayers) [inline test/unit/vdom/shadow-dom.js]: I'd assume existence of `attachShadow`.

File: test/unit/vdom/shadow-dom.js [hunk 5]
```diff
@@ -1,34 +1,23 @@
...
       expect(elem[symbols.shadowRoot]).not.to.equal(elem);
```
Comment (bradleyayers) [inline test/unit/vdom/shadow-dom.js]: Nuke `symbols.shadowRoot`

File: src/api/component.js [hunk 6]
```diff
@@ -257,18 +236,13 @@ Component.prototype = Object.create(HTMLElement.prototype, {
     }
   },
 
-  // v0 and v1
+  // v1
```
Comment (treshugart) [inline src/api/component.js]: Ahh yep, should probably convert that to `// spec` maybe? It's nice to differentiate in the code what is Skate and what is spec'd. What should we do when v2 comes? Will possibly want to differentiate between those, in that case this is future-proof :D

File: src/api/define.js [hunk 7]
```diff
@@ -107,42 +106,22 @@ function prepareForRegistration (name, Ctor) {
...
+    throw new Error('Skate requires native custom element support or a polyfill.');
   }
 
-  throw new Error('Skate requires native custom element support or a polyfill.');
+  let uniqueName = name;
+  if (window.customElements.get(name)) {
+    uniqueName = generateUniqueName(name);
+  }
+  prepareForRegistration(uniqueName, Ctor);
```
Comment (nadiam84) [inline src/api/define.js]: No need in this function anymore, you can just move those two lines here.

File: src/api/define.js (hunk 3 above)
Comment (treshugart) [inline src/api/define.js]: Do you think it's worth keeping here to enable async loading of polyfill / Skate?

File: test/unit/vdom/elements.js [hunk 8]
```diff
@@ -136,49 +135,6 @@ describe('vdom/elements', () => {
...
-        function assertSlotElement () {
-          expect(ch1.tagName).to.equal('SLOT', 'vdom');
-          expect(ch1.getAttribute('name')).to.equal('test', 'vdom');
-          expect(ch2.tagName).to.equal('SLOT', 'vdom.element(slot)');
-          expect(ch2.getAttribute('name')).to.equal('test', 'vdom.element(slot)');
-        }
-
-        if (shadowDomV1) {
-          assertSlotElement();
-        } else if (shadowDomV0) {
```
Comment (treshugart) [inline test/unit/vdom/elements.js]: We've removed support for no shadow DOM, so the only thing we need to test now is SD v1.

File: src/api/define.js (hunk 7 above)
Comment (treshugart) [inline src/api/define.js]: I might keep this for now if that's okay? I'm shuffling this around in #873 anyways and this is less churn to merge.

File: src/api/define.js (hunk 3 above)
Comment (bradleyayers) [inline src/api/define.js]: I meant moving it to line 113 (not the top of the file).

File: test/unit/vdom/elements.js (hunk 8 above)
Comment (bradleyayers) [inline test/unit/vdom/elements.js]: Isn't that what line 168 was doing?

**Pull request merged**